subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -103,6 +105,23 @@ protected SnapshotHierarchy updateNotifyingListeners(UpdateFunction updateFuncti
         }
     }
 
+    @Override
+    public void registerChangeBroadcaster(FileWatcherRegistry.ChangeHandler changeBroadcaster) {
+        if (this.changeBroadcaster != null) {
+            throw new IllegalStateException("A change broadcaster has already been registered: " + changeBroadcaster);
+        }
+        this.changeBroadcaster = changeBroadcaster;
+    }
+
+    @Override
+    public void unregisterChangeBroadcaster(FileWatcherRegistry.ChangeHandler changeBroadcaster) {
+        if (this.changeBroadcaster == changeBroadcaster) {
+            this.changeBroadcaster = null;
+        } else {
+            throw new IllegalStateException("Cannot unregister change broadcaster " + changeBroadcaster + " because it is not registered");
+        }
+    }
```

Comment:
I'm wondering how we could do this better than setting mutable state on a service back and forth.

I wish we had a way to handle this via services instead. We should create an instance of `VirtualFileSystemAccessor` or something for every build invocation and pass the right `ChangeHandler` to it. It feels like creating a new scope in each `BuildActionExecutor` to do this should be easy. I wonder if we could take some steps in that direction.

That said, even if we keep this setter approach, I think the parameter to the `unregister` method is unnecessary. Or do you think we are likely to catch some bug via that check? It's called from `finally` block, and all that, so it feels like defensive programming to do the check.

Reply:
I now use a different approach with a separate service to register the listeners.